### PR TITLE
DM-26822: use a loop running in a background thread to read the DDS queue

### PR DESCRIPTION
**Pull request by r-owen**

### CODE REVIEW

File: python/lsst/ts/salobj/domain.py [hunk 1]
```diff
@@ -274,45 +273,44 @@ def remove_salinfo(self, salinfo):
...
+
+        Intended for exit handlers and constructor error handlers.
+        """
+        try:
+            while self._salinfo_set:
+                salinfo = self._salinfo_set.pop()
+                salinfo.basic_close()
+        finally:
+            self.participant.close()
+        self.participant.close()
```
Comment (tribeiro) [inline python/lsst/ts/salobj/domain.py:287]: Isn't this calling `self.participant.close()` twice? once in `finally` and another here if there's no exception? You can probably remove this one no?

File: python/lsst/ts/salobj/domain.py [hunk 2]
```diff
@@ -274,45 +273,44 @@ def remove_salinfo(self, salinfo):
...
             await self.done_task
             return
+        self.isopen = False
         try:
             while self._salinfo_set:
                 salinfo = self._salinfo_set.pop()
                 await salinfo.close()
         finally:
-            self.close_dds()
+            self.participant.close()
```
Comment (tribeiro) [inline python/lsst/ts/salobj/domain.py:304]: This bit of code from line 298 to here is exactly like the one in `basic_close`, consider calling it instead (if what you want to do is a "basic_close" here.

File: python/lsst/ts/salobj/sal_info.py [hunk 3]
```diff
@@ -720,36 +740,45 @@ async def _read_loop(self):
...
-                        reader.dds_queue_length_checker.check_nitems(len(data_list))
-                        sd_list = [
-                            self._sample_to_data(sd, si)
-                            for sd, si in data_list
-                            if si.valid_data
-                        ]
-                        if sd_list:
-                            reader._queue_data(sd_list)
-                        await asyncio.sleep(0)  # free the event loop
+                await loop.run_in_executor(pool, self._read_loop_thread, loop)
```
Comment (tribeiro) [inline python/lsst/ts/salobj/sal_info.py:743]: Nice!

**Review by tribeiro — COMMENTED** [review]
This is funny. While I was looking into the problem of the producer getting behind last week, I looked at this code and thought you could implement this exactly how you implemented it here. 

I have one request though. I don't know if something changed in `ipython` or in python itself but I can't start remotes on the interactive shell anymore. I get a:

```
/opt/lsst/software/stack/stack/tssw/ts_salobj/python/lsst/ts/salobj/topics/read_topic.py in callback(self, func)
    404         self._data_queue.clear()
    405         self._callback = func
--> 406         self._callback_loop_task = asyncio.create_task(self._callback_loop())
    407
    408     @property

/opt/lsst/software/stack/conda/miniconda3-py37_4.8.2/envs/lsst-scipipe-1a1d771/lib/python3.7/asyncio/tasks.py in create_task(coro)
    348     Return a Task object.
    349     """
--> 350     loop = events.get_running_loop()
    351     return loop.create_task(coro)
    352

RuntimeError: no running event loop
```

exception. If you replace `asyncio.create_task`, by `asyncio.get_event_loop().create_task` in 

```
ts_salobj/python/lsst/ts/salobj/topics/read_topic.py

--> 406         self._callback_loop_task = asyncio.create_task(self._callback_loop())

```

and in 

```
ts_salobj/python/lsst/ts/salobj/remote.py

--> 190                 self.start_task = asyncio.create_task(self.start())
```

it will work just fine.

File: python/lsst/ts/salobj/domain.py (hunk 1 above)
Comment (r-owen) [reply python/lsst/ts/salobj/domain.py:287]: Good catch. I'll delete the second one.

File: python/lsst/ts/salobj/domain.py (hunk 2 above)
Comment (r-owen) [reply python/lsst/ts/salobj/domain.py:304]: Good suggestion.

**Comment by r-owen** [comment]
I made a small change that might fix the event loop issue you were having. Please try again.

**Review by tribeiro — APPROVED** [review]

File: python/lsst/ts/salobj/sal_info.py (hunk 3 above)
Comment (r-owen) [reply python/lsst/ts/salobj/sal_info.py:748]: Thank you.

File: python/lsst/ts/salobj/domain.py (hunk 2 above)
Comment (r-owen) [reply python/lsst/ts/salobj/domain.py:302]: It is not quite the same: this `Domain.close` code calls `await salinfo.close`, whereas `Domain.basic_close` calls `salinfo.basic_close()` synchronously. `SalInfo.close` does a bit more.

**Pull request merged**